app/controllers/api/v1/release_engines/rubygems/compact_index_controller.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+require 'compact_index'
+
+module Api::V1::ReleaseEngines
+  class Rubygems::CompactIndexController < Api::V1::BaseController
+    before_action :scope_to_current_account!
+    before_action :require_active_subscription!
+    before_action :authenticate_with_token
+    before_action :set_packages, only: %i[versions names]
+    before_action :set_package, only: %i[info]
+
+    def ping
+      skip_verify_authorized!
+
+      head :ok
+    end
+
+    def versions
+      authorize! packages,
+        to: :index?
+
+      artifacts = authorized_scope(current_account.release_artifacts.unyanked.for_packages(packages.ids)).preload(:specification, :package, release: %i[product entitlements constraints])
+      authorize! artifacts,
+        to: :index?
+
+      gems = artifacts.group_by(&:package).reduce([]) do |arr, (package, artifacts)|
+        versions = to_versions(artifacts)
+
+        arr << CompactIndex::Gem.new(package.key, versions.sort)
+      end
+
+      render plain: CompactIndex.versions(
```

Comment:
Should we add etag support? Looks like gem internals use it for caching purposes to reduce load.